doc/writing_new_controller.rst

```diff
@@ -0,0 +1,138 @@
+   Use the ``--help`` flag for more information on how to use it.
+   There is also an option to create library source files and compile rules to help you in the following steps.
+
+2. **Preparing source files**
+
+   After creating the package, you should have at least ``CMakeLists.txt`` and ``package.xml`` files in it.
+   Create also ``include/<PACKAGE_NAME>/`` and ``src`` folders if they do not already exist.
+   In ``include/<PACKAGE_NAME>/`` folder add ``<contorller_name>.hpp`` and ``<controller_name>.cpp`` in the ``src`` folder.
+   Optionally add ``visibility_control.h`` with the definition of export rules for Windows.
+   You can copy this file from an existing controller package and change the name prefix to the ``<PACKAGE_NAME>``.
+
+3. **Adding declarations into header file (.hpp)**
+
+   1. Take care that you use header guards. ROS2-style is using ``#ifndef`` and ``#define`` preprocessor directives. (For more information on this, a search engine is your friend :) ).
+
+   2. include ``"controller_interface/controller_interface.hpp"`` and ``visibility_control.h`` if you are using one.
+
+   3. Define namespace for your controller. This is usually a package name written in ``snake_case``.
+
+   4. Define the class of the controller, extending ``ControllerInterface``, e.g.,
+      .. code:: c++
+         class ControllerName : public controller_interface::ControllerInterface
+
+   5. Add constructor without parameters and the following public methods overriding ``ControllerInterface`` definition: ``init``, ``command_interface_configuration``, ``state_interface_configuration``, ``on_configure``, ``on_activate``, ``on_deactivate``, ``update``.
+      For exact definitions check the ``controller_interface/controller_interface.hpp`` header or one of the controller from `ros2_controllers <https://github.com/ros-controls/ros2_controllers> `_.
+
+   6. (optional) Often, controllers accept lists of joint names and interface names as parameters.
+      If so, you can add two protected string vectors where to store those values.
+
+4. **Adding definitions into source file (.cpp)**
+
+   1. Include the header file of your controller and add a namespace definition to simplify further development.
+
+   2. (optional) Implement constructor if needed. There, you could initialize member variables.
+      This could also be done in the ``init`` method.
+
+   3. Implement the ``init`` method. The first line usually calls for the parent ``init`` method.
+      Here is the best place to initialize the variables would be best if you initialize all the variables, reserve memory, and most important, declare node parameters used by the controller. If everything works fine return ``controller_interface::return_type::OK`` or ``controller_interface::return_type::ERROR`` otherwise.
+
+   4. Write the ``on_configure`` method. Parameters are usually read here, and everything is prepared so that controller can be started.
```

Comment:
```suggestion
   4. Write the ``on_configure`` method. Parameters are usually read here, and everything is prepared so that the controller can be started.
```